[test] dynamo: dont hold onto real tensors in the __dict__ of any graph inputs

bdhirsh opened the pull request:
This seems like it should fix https://github.com/pytorch/pytorch/issues/132290, although we probably need a better strategy (comments below)

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #132338
* #132524
* #132322
* #132337
* #132243



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

bdhirsh committed:
dynamo: dont hold onto real tensors in the __dict__ of any graph inputs

[ghstack-poisoned]

bdhirsh commented on torch/_dynamo/variables/builder.py:
```diff
@@ -1497,7 +1497,7 @@ def wrap_tensor(self, value: torch.Tensor):
         self.tx.output.input_source_to_var[source] = tensor_variable
         assert "tensor_dict" not in tensor_proxy.node.meta
-        tensor_proxy.node.meta["tensor_dict"] = value.__dict__.copy()
+        tensor_proxy.node.meta["tensor_dict"] = {k: None if isinstance(v, torch.Tensor) else v for k, v in value.__dict__.items()}
```
two things from talking to Joel:

(1) we should probably be using weakrefs to hold onto the tensors (right now I just hackily null out any tensor attributes to avoid a leak, although I don't know if dynamo actually relies on them being populated here)

(2) this doesn't handle attributes with further nested that hold tensors.

albanD reviewed (commented):
This will fix the issue for sure but not sure what this field is used for today?
if this is used for debugging, you could put in that field a special marker that says that this was a Tensor (as opposed to afield with None).

jbschlosser commented on torch/_dynamo/variables/builder.py:
```diff
@@ -1497,7 +1497,7 @@ def wrap_tensor(self, value: torch.Tensor):
         self.tx.output.input_source_to_var[source] = tensor_variable
         assert "tensor_dict" not in tensor_proxy.node.meta
-        tensor_proxy.node.meta["tensor_dict"] = value.__dict__.copy()
+        tensor_proxy.node.meta["tensor_dict"] = {k: None if isinstance(v, torch.Tensor) else v for k, v in value.__dict__.items()}
```
is a similar change needed here?
https://github.com/pytorch/pytorch/blob/f0da167ce55ddbbe4f0138ed0f0804008069ddad/torch/_dynamo/output_graph.py#L810

jbschlosser replied:
oh and here:

https://github.com/pytorch/pytorch/blob/6b28af1b79eaa63e2f423d925bbd42330582983f/torch/_dynamo/variables/builder.py#L1141-L1143

soulitzer commented:
From the objgraph in the issue it looks like the refcycle is entirely in python. Curious if we would expect gc to have cleared this cycle? Locally it appears that the reference to the inner tensor gets cleared by  `torch._dynamo.reset()` + `gc.collect()` only when the inner tensors are on cpu, on cuda however reset + gc seems to be not enough. Curious what is different for cuda...

ezyang commented:
My hottest of hot takes is that `tensor_dict` has no business being on meta at all.

bdhirsh commented:
> Curious if we would expect gc to have cleared this cycle? Locally it appears that the reference to the inner tensor gets cleared by torch._dynamo.reset() + gc.collect() only when the inner tensors are on cpu, on cuda however reset + gc seems to be not enough. Curious what is different for cuda...

yep, locally for me, a combo of `torch._dynamo.reset()` + `gc.collect()` freed the memory. That does not seem ideal though

> My hottest of hot takes is that tensor_dict has no business being on meta at all.

I put this PR up mostly for discussion, since i'm not too familiar with `tensor_dict` (cc @mlazos )

bdhirsh committed:
Update on "[test] dynamo: dont hold onto real tensors in the __dict__ of any graph inputs"

This seems like it should fix https://github.com/pytorch/pytorch/issues/132290, although we probably need a better strategy (comments below)




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bdhirsh committed:
Update on "[test] dynamo: dont hold onto real tensors in the __dict__ of any graph inputs"

This seems like it should fix https://github.com/pytorch/pytorch/issues/132290, although we probably need a better strategy (comments below)




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bdhirsh committed:
Update on "[test] dynamo: dont hold onto real tensors in the __dict__ of any graph inputs"

This seems like it should fix https://github.com/pytorch/pytorch/issues/132290, although we probably need a better strategy (comments below)




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bdhirsh committed:
Update on "[test] dynamo: dont hold onto real tensors in the __dict__ of any graph inputs"

This seems like it should fix https://github.com/pytorch/pytorch/issues/132290, although we probably need a better strategy (comments below)




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

mlazos commented:
Hi, I'm not sure what it was originally used for but the most important thing that I know it's used for today is passing static parameters/buffer information through the stack to cuda graphs. This information is only carried by a string flag so removing tensors from this should be fine. I will review this PR later today.

mlazos commented on torch/_dynamo/variables/builder.py:
```diff
@@ -1506,7 +1506,7 @@ def wrap_tensor(self, value: torch.Tensor):
         self.tx.output.input_source_to_var[source] = tensor_variable
         assert "tensor_dict" not in tensor_proxy.node.meta
-        tensor_proxy.node.meta["tensor_dict"] = value.__dict__.copy()
+        tensor_proxy.node.meta["tensor_dict"] = {k: None if isinstance(v, torch.Tensor) else v for k, v in value.__dict__.items()}
```
you should extract this into a utility function, `tensor_dict` is populated in multiple places

bdhirsh committed:
Update on "[test] dynamo: dont hold onto real tensors in the __dict__ of any graph inputs"

This seems like it should fix https://github.com/pytorch/pytorch/issues/132290, although we probably need a better strategy (comments below)




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

bdhirsh commented:
Closed in favor of https://github.com/pytorch/pytorch/pull/132805
